packages/flutter/lib/src/widgets/app_model.dart

```diff
@@ -0,0 +1,164 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import 'framework.dart';
+import 'inherited_model.dart';
+
+/// Enables sharing key/value data with the all of the widgets below [child].
+///
+/// - `AppModel.set(context, key, value)` changes the value of an entry
+/// in the shared data table and forces widgets that depend on that entry
+/// to be rebuilt.
+///
+/// - `AppModel.get(context, key)` creates a dependency on the key and
+/// returns the value for the key from the shared data table.
+///
+/// - `AppModel.init(context, key, value)` changes the value of an entry
```

Comment:
This still strikes me as a source of hard-to-debug bugs. Can we assert that nobody has started depending on the value stored under `key` when `init` is called to enforce the desired contract?

Alternatively, maybe `getValue` should take an initializer closure that executes if no value has been stored under `key` so far and puts its return value into the map. It would presumably be a required param to guarantee that the first one to ask for a value initializes it and all others asking for the same value then see that one.

Reply:
AppModel.get() now requires a callback parameter which is used to create the initial value for a keyword.